### ISSUE TITLE
Osara Chord Navigation Does Not Respect Item Split Points

### ISSUE THREAD

**Issue by MatejGolian** [jcsteh/osara#727]
Hello, It's me again... :-)
It's best for me to explain this on an example.
I've attached a tiny sample project below. Just the C major scale entered in 2 octaves across 2 bars, so it's 16 MIDI notes in total.
First I had everything as one item. Than I made a split at bar 2. The problem is that when I open the second item and land on position 2.1.00, I can still move to the notes that aren't audible in that particular item with Osara's Chord navigation. So, pressing the left arrow at 2.1.00 will first move to the note at 1.4.50, than 1.4.00 and so on.
Technically Osara's not wrong, because after the split the second item does indeed still contain all the notes. That's easily confirmed by opening the event list. Oddly enough, the first item only has 8 notes after the split. Of course I know that it all comes down to how Reaper handles splitting MIDI items, but this behavior is quite interesting nevertheless...
Anyway, Selecting all notes with CTRL + A results in 8 notes being reported in case of both items.
So, my question is, wouldn't it be better if Osara's chord navigation and note selection commands only operated within the active range of a given MIDI item in cases like this?
The point could be made that the current behavior is logical and understandably so. To me it still kinda feels odd in practice though. I ultimately think however, that the best solution would be to make sure Osara's behavior mimicks the user experience of sighted users. If they can see and go beyond the split point in the piano roll as well, the current behavior is OK, but if they can't do so, IMO it would make more sense if Osara honored the split points. What do you think?
[Test.rpp.txt](https://github.com/jcsteh/osara/files/8310824/Test.rpp.txt)

**Comment by jcsteh** [comment]
Yeah, I find this behaviour deeply frustrating myself. I always thought this was just how REAPER handled things. However, if I understand you correctly that select all only selects the notes within the item's range, that suggests the visual behaviour might indeed restrict. Would anyone be able to get a working pair of eyeballs looking at this?

**Comment by MatejGolian** [comment]
Yes, select all does only operate within the item's range. BTW, I tested this by selecting all notes and than hitting delete as well and only the notes that were within that range got deleted, but I could still navigate to the notes beyond the split point. I haven't tested how CC lanes behave in this regard, but in case this note thing gets addressed, maybe the CC stuff will need to be fixed as well. My hunch is that sighted users only see the active part, but of course, we'll see once someone takes a look at it...

**Comment by MatejGolian** [comment]
I've checked one more thing I haven't thought of before. Going to the start of the file with W or CTRL + Home ignores the inactive notes too. Since the action is literally called go to start of file, it would be really strange, at least from a linguistic perspective, if sighted people could still see past the split point, but of course one never knows...

**Comment by MatejGolian** [comment]
@AZSlow3, in case you get this message, since you're a sighted person, would you mind taking a look at how it looks visually in the MIDI editor? Your help would be greatly appreciated.

**Comment by AZSlow3** [comment]
I can confirm that OSARA ignores MIDI item boundaries. So OSARA navigates and reports notes which technically do not exist in the project, effectively reporting complete media source ignoring item's start in source and length.

Tested with OSARA 2022.1pre-907 in REAPER 6.49, Windows.

**Comment by MatejGolian** [comment]
Thank you for the info and verification, Alexey!
Honestly, I am glad that your feedback confirmed my assumptions. The current behavior has always been annoying the hell out of me, but like Jammie, I too believed that that's just how things were supposed to work in the piano roll.

**Comment by AZSlow3** [comment]
I have not checked carefully, but so far I have not found any code in midiEditorCommands.cpp which deals with Item position nor length. So I have also checked that OSARA is not navigating throw notes in looped MIDI items belong the source, that also requires calculation on item properties. I mean the whole logic of source to item relation is probably not implemented.

**Comment by jcsteh** [comment]
You're correct; it's not implemented. I had assumed that either REAPER visually displays all notes or that the MIDI API functions would only return notes clipped to item boundaries. Apparently that is not the case. So, OSARA will have to clip itself.

**Comment by MatejGolian** [comment]
May I ask you 2 clever programming guys what you think about the API functions not clipping the notes? My first thought was that it sucks, but than I thought some more and I'd be curious to know what you think. Could the fact that the API functions always return all notes actually be a good thing in a way? Sure, in this particular case it doesn't produce desired results out of the box, but could it be argued that the fact that the API functions are not trying to be too smart makes them more predictable? - Since they always return everything, the programmer is able to do what ever he or she wants with the output at all times. Am I off on this or does this make sense?
And @jcsteh, since Osara will have to clip the notes, won't time signature and tempo changes cause problems? I hope not.May I ask you 2 clever programming guys what you think about the API functions not clipping the notes? My first thought was that it sucks, but than I thought some more and I'd be curious to know what you think. Could the fact that the API functions always return all notes actually be a good thing in a way? Sure, in this particular case it doesn't produce desired results out of the box, but could it be argued that the fact that the API functions are not trying to be too smart makes them more predictable? - Since they always return everything, the programmer is able to do what ever he or she wants with the output at all times. Am I off on this or does this make sense?
And @jcsteh, since Osara will have to clip the notes, won't time signature and tempo changes cause problems? I hope not.

**Comment by AZSlow3** [comment]
Not only clip but also loop. Looped source is infinite in both directions and limited by item size only. 
Partial notes which technically start before item but end inside item start at item boundary. Not sure if playback rate should also be accounted manually. I am not experienced in REAPER API, I had to understand all these relations in project files for my converter from Sonar.

@MatejGolian REAPER API tend to expose the information in the project using internal terms. That is not always intuitive for users. While for OSARA auto-clipping is this case could be useful, for most script developers that will be disaster.

**Comment by MatejGolian** [comment]
I'm not sure I understand all of that, but in that case wouldn't it make sense to ask for a fix from Cockos?

**Comment by AZSlow3** [comment]
In short, there is no problems with API.

**Comment by MatejGolian** [comment]
I see - thank you for the clarification. And if you guys don't mind the reminder, I have checked the CC navigation commands and it seems that they exhibit very similar (or maybe identical behavior). So those will probably need fixing as well. Should I open a new issue for that or will changing the title of this issue suffice? I'd say that getting the note part working correctly takes priority though.

**Comment by jcsteh** [comment]
We'll handle CCs in the same issue. I doubt we're going to get this
entirely correct on the first pass - as noted, there are some obscure edge
cases like notes that overlap item boundaries and those are going to be
tricky - but we should be able to manage something.

**Comment by jcsteh** [comment]
@leonardder, I'm trying to figure out the best way to deal with this. Do you think we should just handle this when returning from findChord or findCC or do you think we should tweak the iterators somehow? Aside from preventing the return of an event outside the item boundaries, one catch that isn't immediately obvious is that we also need to tweak note lengths if they're clipped by the item end boundary. That's going to be messy because length calculation is handled by MidiNote, although I guess getLength could be passed the MIDI take so it can consider the item boundaries.

Doing this in findChord/findCC is probably going to be easier, but it feels weird because the iterator can iterate through the entire source, even though everything we ever do with it should be restricted to the item boundaries. On the other hand, doing it in the iterator is probably going to be very messy. We'd probably have to make the iterators return invalid starts and ends or something to keep the binary search happy. Alternatively, we could fudge the indexes so that they're restricted to item boundaries; i.e. 0 is always the first event in the item, count is always the last. That would require the iterators to keep track of an offset. The potential for confusion and code complexity there is huge, so I'm not too keen on that idea.

**Comment by jcsteh** [comment]
My current thinking is that we handle this in findNote/findCC. We need to tweak both the start and length if a note overlaps the item start, so maybe MidiNote could have a clipToItem method which is passed an item and tweaks start and end appropriately.

**Comment by MatejGolian** [comment]
Thanks @jcsteh. I've edited the issue title accordingly and I have also updated my initial comment by adding a second test file with an overlapping note.
If it's not too complicated, may I ask why is it also necessary to tweak note lengths if the note overlaps the item start boundary? In my new example the overlapping note is entered one beat before the item start boundary, but that note doesn't start playing until the play head reaches the start boundary it self - anything you may hear from before the start boundary of the second item only comes from item number 1. That's pretty obvious, but when I go back and look at that overlapping note, Osara reports the exact same length for it as can be seen in the event properties dialog, so why does the note length need to be recalculated? How should Osara deal with cases like this in theory? Sorry if you already explained this in other terms, but will overlapping notes like this get ignored by Osara (because they start before the item start boundary) or will Osara still pick them up (because they start producing sound once the start boundary is reached)?
I'm just asking - I don't know which behavior is correct and close to the visual UX.

**Comment by MatejGolian** [comment]
I guess I'll answer my own question to a degree.
Performing a select all operation in case there's an overlapping note before the start boundary selects that overlapping note as well, so if I understand the situation, Osara should take these kind of overlapping notes into account as well. And regarding note length recalculations, do I assume that these would only be internal ones and that Osara's speech output would not get affected by them? I ask because to me Osara's current spoken feedback seems absolutely fine when it comes to note lengths.

**Comment by leonardder** [comment]
I'm ridiculously confused.
        
        @AZSlow3 wrote:
        > In short, there is no problems with API.
        
        I'm not convinced of that. Functions like MIDI_GetNote, MIDI_CountEvts, etc. all operate at the take level. So first, we need to know what is meant by a take in the MIDI editor. As far as I know, there is a tree like structure like track > item > take, so there's still the question why a take overlaps multiple items in this case. Or is my assumption wrong and could an item have multiple takes, where a single take could also span multiple items?
        
        @Jcsteh re your questions: I think we should find out first whether REAPER's API is really correct here? OSARA's behavior in the midi editor has been this way for ages, so I don't think we shouldn't rush too much.

**Comment by leonardder** [comment]
Ugh, never mind my previous comment. I was just very confused and after reading the initial message in the thread over and over again, it all became much clearer for me. 

I think the underlying issue is very similar to filtered notes. However, filtered notes are ignored by MIDI_GetNote, where as notes that are across item boundaries are not. So yes, while the API is technically correct, the reason to exclude filtered notes and include notes beyond the item's boundaries is confusing and it certainly makes sense to contact Cockos about this. I'm happy to do this as it is a nice follow up to an earlyer conversation I had about the filtered notes which resulted in #659.

@jcsteh wrote:
> @leonardder, I'm trying to figure out the best way to deal with this. Do you think we should just handle this when returning from findChord or findCC or do you think we should tweak the iterators somehow?

While I think tweaking the iterator might be messy, how about the following? We could add two methods to the iterator, moveToStartOfTake/moveToEndOfTake that has all the logic in place to position the iterator at the respective  item's boundary. Then, in findChord/findCC, we can swap moveToEnd with moveToEndOfTake and position the begin iterator at the start of the take with moveToStartOfTake. Of course these methods will have to iterate themselves which might be somewhat slower than handling everything in findCC/findChord, but the find methods are already pretty complex and will become much more complex when we'd have to handle boundaries there.

**Comment by AZSlow3** [comment]
@leonardder  Facts. An item can have several takes. Only one of that takes can be current at any particular time and it is displayed in MIDI editor, while there is a flag to play all takes. So other takes can be audible while not appearing in MIDI editor, that is not a bug. Each take has own source and associated playback rate, source offsets and some other parameters. Item position, length and source looping properties are item wide. 
The following is just by guess. When asked about takes, REAPER API works with take only. So it does not apply Item properties on takes, like position, length and looping. So corresponding calls are not providing the information currently visible in MIDI Editor, but the information about take as in the project data.

It can be valid request to ask Cockos provide jet another API, which return MIDI Editor information about takes, so after Item properties applied. Depending from real structure and logic in the MIDI Editor, it can be easy or hard to export that.

**Comment by leonardder** [comment]
Thanks for clarifying!
To put it the other way around, if you have an item and split that halfway, isn't it just weird that the first item contains all events from start to split and the second item contains events from start to end, where the item's boundary is at split so everything before split isn't played, yet part of the item? Isn't that just a bug to begin with?

**Comment by jcsteh** [comment]
Yeah, that inconsistency confuses me. I could maybe understand if both takes included all data, since the raw data has to be present underneath to allow for growing the item. But that logic falls down because the first item doesn't include the data after the split.

**Comment by MatejGolian** [comment]
That aspect of MIDI editing in Reaper feels buggy indeed. If all MIDI items contained all data after split operations, it could be argued that the point is to make MIDI editing non-destructive, but as it is now, it's really just confusing.
In layman's terms, assuming the Reaper API doesn't change, Osara's task should be just to ignore the notes beginning before the item start boundary which do not continue on into the active part of the item. Of course, I understand that this is easier said than done. In any case, Osara should only provide access to those events that get affected by the select all operation - nothing less, nothing more.

**Comment by AZSlow3** [comment]
I think it make sense distinguish between REAPER behavior and how OSARA represent something.
For splitting, REAPER probably has good reasons to trim or not to trim in some situations. There is "Allow trim of MIDI items on splitting" option, so user can forbid trimming. The option is not "do trim", in other words it does not trim everything. In practice when allowed, REAPER trims both sides of not looped item and left side of looped MIDI item.
All that should not influence OSARA. When trimmed, takes are reported as such.

**Comment by leonardder** [comment]
I wonder whether we shouldn't just start with reporting notes that are fully out of the visible part of the take as clipped and partially visible notes as cropped.

**Comment by MatejGolian** [comment]
I'd personally prefer if Osara's event navigation ignored the events that are fully outside of the visible part of the take altogether when it comes to the piano roll. I think this is the only thing that maybe needed to make things behave as people would generally expect. I'm guessing that as long as it's possible to identify these 'clipped' notes, it should also be possible to make Osara ignore them. Or am I mistaken?
Reporting if a given note is cropped would be a nice addition and could help clarify the UX.

**Comment by AZSlow3** [comment]
I have tried to manually "glue" MIDI item into new track in Lua, scripting is way faster then playing inside OSARA.
With attached script I have checked it works under many border conditions like changed play rate, looping, offset
and fixed tempo in the source.
Some calculations was not obvious at the beginning. SWS use different approach to find take source length in PPQ,
but my version does not use SWS and works correctly so far.

For OSARA, we need to define what to do with looped items. Visually they are represented as glues, so containing repeated events, but with markers at the places of loop boundary. But there is one technical problem with such approach. Visually once one note is selected, all instances of this note are selected. So there is no sense of particular time and in API reported selected note is always in the first loop, potentially inside trimmed part of it. In OSARA event time make sense, at least to define which notes are neighbors.

We can allow editing just one loop. In case the beginning is trimmed we still need provide access to that trimmed part. These notes should be edited inside the second loop instance, if item has sufficient length.

In both cases it can happened the same note technically has up to 3 different lengths. For example when the first note is partially trimmed, exist as complete in the next loop and trimmed right at the end of the item. 

[midi_glue.zip](https://github.com/jcsteh/osara/files/8621938/midi_glue.zip)

**Comment by MatejGolian** [comment]
I haven't used looping MIDI items yet (or any kind of looping for that matter), so I don't fully understand the issue here, but I see you know what you're talking about. :D I can't give you any advice, but thanks for working on this...

**Comment by jcsteh** [comment]
Let's say your item starts at beat 4, but the note starts at beat 3 and is
2 beats in length. When you cursor over it with OSARA, I imagined it should
position the cursor at beat 4 (not beat 3) and report/preview the note as
being 1 beat in length, since we're restricting to item boundaries.
However, you could argue it should still position at beat 3 and report as 2
beats as it does now. It's a tricky one. It feels weird to me to restrict
to item boundaries for navigation but still allow the cursor to drift
outside those boundaries for overlaps. On the other hand, I don't like the
idea of lying to the user either.

It'd be interesting to know what happens visually here. If REAPER only
shows notes inside item boundaries, how does it show notes that overlap the
start boundary?

**Comment by AZSlow3** [comment]
REAPER shows notes inside item boundaries. In your example, the note visually starts at beat 4. As I wrote before, visual representation is for glued and so trimmed item, just with extra visual markers at loop points.

But editing is different. All looped instances of the same note are visually edited in sync. Dragging outside of item borders is allowed, as long as the source has space, but the result is not shown outside the item. That is true for all items and operations. For example select by pitch selects trimmed notes as well and future manipulations affect them, even so there is no visual confirmation that is happening.

My proposal is still the same. In any case allow editing inside item boundary only. For trimmed part at the beginning, allow editing it in the first copy, as long as the item boundary include corresponding notes. So it can happened the same note can be edited at 2 different places, reporting 2 different boundaries. Looping point should be explicitly announced. I think such approach has closest match to visual operations.

Since in looped item the source clip is trimmed at looping point by design, there can be no notes which cross looping point. But since two instances of the same note can be edited separately and REAPER has no such splitting, the information which version is currently edited has to be preserved in OSARA.

**Comment by MatejGolian** [comment]
So how difficult is it to filter out trimmed events that don't last beyond the start boundary? And conversely, how difficult is it to determine that a given trimmed note crosses the start point?
Isn't this one of those things that could be negatively impacted by features affecting time, such as tempo/time signature markers or stretch markers?
